### PR TITLE
Fixed #34789 -- Avoided filter_horizontal misbehaving when instance of same model is added via JS add action.

### PR DESCRIPTION
**Pull request by yokeshwaran1**
Updated query selectors to targe source options which will prevent duplication in "Chosen" column.
Ticket Link: https://code.djangoproject.com/ticket/34789#comment:8

Screenshot 2023-09-02 at 2 46 09 AM
Screenshot 2023-09-02 at 2 46 29 AM

### CODE REVIEW

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js [hunk 1]
```diff
@@ -92,12 +92,7 @@
         // select (+ or :pencil:) update ForeignKey PK of the rest of selects
         // in the page.
 
-        const path = win.location.pathname;
-        // Extract the model from the popup url '.../<model>/add/' or
-        // '.../<model>/<id>/change/' depending the action (add or change).
-        const modelName = path.split('/')[path.split('/').length - (objId ? 4 : 3)];
-        // Exclude autocomplete selects.
-        const selectsRelated = document.querySelectorAll(`[data-model-ref="${modelName}"] select:not(.admin-autocomplete)`);
+        const selectsRelated = document.querySelectorAll(`[data-context="available-source"]`);
```
Comment (nessita) [inline django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js:95]: I don't understand why we can drop the `data-model-ref="${modelName}"` here, what if there are multiple fields in the same form that allow for the multiple selection process?

**Review by nessita — COMMENTED** [review]
Added a question, thanks!

File: django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js (hunk 1 above)
Comment (yokesh440) [reply django/contrib/admin/static/admin/js/admin/RelatedObjectLookups.js]: Hi @nessita , thanks for ur feedback. Valid point. My intention behind removing the data-model-ref="${modelName}" was to simplify the selector logic, but I understand the need for specificity to make sure we r targeting correct model. I have reintroduced the data-model-ref="${modelName}".

**Comment by yokeshwaran1** [comment]
Thank @nessita for pointing out the additional issue. I will reproduce it and create a separate ticket as u suggested. appreciate ur feedback, guidance

**Comment by nessita** [comment]
> Thanks @nessita for pointing out the additional issue. I will reproduce it and create a separate ticket as u suggested. appreciate ur feedback, guidance

Thank you @yokesh440 for your quick response!

Separately, I believe that the common practice for code changes to this javascript code is to add a few selenium tests ensuring the correctness of the changes. Would you be able to add some of those tests?

See as a reference the specific file: https://github.com/django/django/blob/main/tests/admin_views/test_related_object_lookups.py

**Comment by yokeshwaran1** [comment]
> > Thanks @nessita for pointing out the additional issue. I will reproduce it and create a separate ticket as u suggested. appreciate ur feedback, guidance
> 
> Thank you @yokesh440 for your quick response!
> 
> Separately, I believe that the common practice for code changes to this javascript code is to add a few selenium tests ensuring the correctness of the changes. Would you be able to add some of those tests?
> 
> See as a reference the specific file: https://github.com/django/django/blob/main/tests/admin_views/test_related_object_lookups.py

Thanks for the feedback. I understand the importance. I will get started with Selenium tests for my changes.

**Comment by yokeshwaran1** [comment]
Hi @nessita 
I have added the Selenium test for the same. Could you please review, merge the PR when you get a chance?

**Comment by nessita** [comment]
> Hi @nessita I have added the Selenium test for the same. Could you please review, merge the PR when you get a chance?

Thank you! I appreciate the effort. The branch is looking good, great work so far.

There is still an issue to resolve: the selenium test you added seems to only assert that the `data-context` attribute is set, which is something that was already tested via the existing tests that you modified (the ones that assert over resulting HTML).

What is needed instead of that, is one or more tests that exercise the user flow in the admin screen, simulating the clicks that an admin user would perform when doing the operations to add an instance via the JS add button, and then asserting that the right values appear in the right places, and that no unexpected value appears in the `Chosen` column.

So basically the new selenium test(s) would mimic the actual steps that a real admin user would take to trigger the bug (but then it will assert the code is actually behaving properly). Does that make sense?

**Comment by nessita** [comment]
Hello @yokeshwaran1! Would you have time to keep working on this? Thank you!

**Comment by yokeshwaran1** [comment]
> Hello @yokeshwaran1! Would you have time to keep working on this? Thank you!

Hey @nessita, Im learning Selenium Automation which is new to me.  So, facing basic issues which Im working to understand & fix.   Shall we pls have the automation ticket as a separate one?

**Comment by nessita** [comment]
> > Hello @yokeshwaran1! Would you have time to keep working on this? Thank you!
> 
> Hey @nessita, Im learning Selenium Automation which is new to me. So, facing basic issues which Im working to understand & fix. Shall we pls have the automation ticket as a separate one?

Hello @yokeshwaran1! Thank you for keep working on this.

We can't merge this work without the corresponding tests for the new logic, so we will need to wait for the selenium tests before being able to land this work.

We can wait, there is no rush, feel free to take your time while you learn the new concepts. Also, feel free to reach out for help in the Django Forum or in the Django discourse if you need to.

**Comment by nessita** [comment]
Hello @yokeshwaran1! Were you able to make any progress with the tests? Thanks!

**Comment by yokeshwaran1** [comment]
> Hello @yokeshwaran1! Were you able to make any progress with the tests? Thanks!

Hello @nessita 
Thank u for checking in. I've been working on the Selenium tests for the "Transitions" section in the Django admin interface. I have encountered an issue where there s an inconsistency in the admin interface's display—specifically, the "Transitions" section appears in one scenario(attached screenshot)

Screenshot 2023-12-19 at 1 03 09 PM

 but Transitions is absent in another during automated testing. (attached screenshot for same)

Screenshot 2023-12-19 at 1 02 58 PM

Im trying to pinpoint the root cause of ths by checking any Settings or configs.  Given the complexity of this, I wonder if u might have encountered something similar before or could provide any insights that might assist in its resolution.

I apologize for delay in Selenium automation due to these unexpected challenges Im facing & im new to automtation. Im prioritizing a resolution to this matter & appreciate any guidance u can offer in the interim.

**Comment by nessita** [comment]
> Im trying to pinpoint the root cause of ths by checking any Settings or configs. Given the complexity of this, I wonder if u might have encountered something similar before or could provide any insights that might assist in its resolution.

@yokeshwaran1 Could you please push what you have built so far? What you describe sounds like your model is not registered with the admin tests. Note that the models that are available to the admin Selenium tests are those described in `django/tests/admin_views/models.py`.  There is no model named `Transition` in that file, so my advice would be to use a model from that file. You can see examples of existing tests in the file `django/tests/admin_views/tests.py`.

> I apologize for delay in Selenium automation due to these unexpected challenges Im facing & im new to automtation. Im prioritizing a resolution to this matter & appreciate any guidance u can offer in the interim.

You can take as much time as you need, we ask from time to time to contributors if they'd have the time to keep working on branches to close those that are stale.

Thank you!

**Comment by devin13cox** [comment]
Hi @yokeshwaran1! I just wanted to check in as I initially posted this ticket. Are you still working on the Selenium tests or were you hoping to pass those off? I'm not familiar with Selenium myself but I can take a stab at it if this has stalled. Just let me know how you want to move forward. Thanks!

**Comment by nessita** [comment]
Hello @devin13cox, thank you for your interest in helping Django! If you have time to work on this, please do so since the original author does not seem to have time to do so (and that's fine, everyone contributes what/when they can).

I'll close this PR, please assign the ticket to yourself. You can base your work on top of this branch, just credit the author using the `Co-authored-by` feature in the commit. Let me know if you have any questions!

**Comment by devin13cox** [comment]
> Hello @devin13cox, thank you for your interest in helping Django! If you have time to work on this, please do so since the original author does not seem to have time to do so (and that's fine, everyone contributes what/when they can).
> 
> I'll close this PR, please assign the ticket to yourself. You can base your work on top of this branch, just credit the author using the `Co-authored-by` feature in the commit. Let me know if you have any questions!

Hey @nessita , had some time this morning so built out the automation tests. Ran into a small issue but I don't think it's a testing problem--it's either an issue that already existed or a new one from this change. Essentially, when a new item is added, this solution correctly prevents it from being added to the "Chosen" column, however it also prevents it from being added to the "Available" column. I would assume that we still want it showing in the "Available" column.

I also saw above a mention of a bug that also existed in `main`. Is this what you were referencing? I haven't been able to track down where we could add the new `data-context="available-source"` attr yet to fix this in the case that it's a new issue, so if you happen to know where this lives I would appreciate direction. 

I can push a PR whenever (although I want to make sure I do it the right way as this is my first django ticket).
